investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace TRINV.Application.Queries.Dashboard;
+
+using DigitalCurrency;
+using Domain.Entities;
+using Domain.Enums;
+using Interfaces;
+using MediatR;
+using Shared.Business.Exceptions;
+using Shared.Business.Extension;
+using Shared.Business.Utilities;
+
+public record GetInvestmentPerformanceQuery(InvestmentType InvestmentType) : IRequest<OperationResult<List<InvestmentPerformanceQueryModel>>>;
+
+
+internal class GetInvestmentPerformanceQueryHandler : IRequestHandler<GetInvestmentPerformanceQuery, OperationResult<List<InvestmentPerformanceQueryModel>>>
+{
+    readonly IRepository<Investment> _investmentRepository;
+    readonly IUserContext _userContext;
+    readonly IMediator _mediator;
+
+    public GetInvestmentPerformanceQueryHandler(IRepository<Investment> investmentRepository, IUserContext userContext, IMediator mediator)
+    {
+        _investmentRepository = investmentRepository;
+        _userContext = userContext;
+        _mediator = mediator;
+    }
+
+    public async Task<OperationResult<List<InvestmentPerformanceQueryModel>>> Handle(GetInvestmentPerformanceQuery request, CancellationToken cancellationToken)
+    {
+        var operationResult = new OperationResult<List<InvestmentPerformanceQueryModel>>();
+
+        // TODO: This needs to be removed when stocks API's are added.
+        if (request.InvestmentType != InvestmentType.Cryptocurrency)
+            return operationResult.ReturnWithErrorMessage(new NotFoundException());
+
+        var currentUserInvestments = await _investmentRepository
+            .GetAllWithPredicateAsync(i => i.UserId == _userContext.UserId
+                                           && i.InvestmentType == request.InvestmentType, cancellationToken);
+
+        var groupedInvestments = currentUserInvestments
+            .GroupBy(i => i.AssetId)
+            .Select(x => new
+            {
+                Symbol = x.Key,
+                Name = x.First().Name,
+                TotalQuantity = x.Sum(i => i.Quantity),
+                TotalValueInFiat = x.Sum(i => i.PurchasePrice),
+                AverageValuePerUnitInFiat = x.Sum(i => i.PurchasePrice) / x.Sum(q => q.Quantity),
+            });
+
+        Dictionary<string, decimal> currentPrices = new Dictionary<string, decimal>();
+
+        if(request.InvestmentType == InvestmentType.Cryptocurrency)
+        {
+            var currentCryptoPrices = await _mediator.Send(new GetDigitalCurrencyListQuery(), cancellationToken);
+
+            foreach (var item in groupedInvestments)
```

Comment:
Here we can remove the foreach and do it something like ''

 currentPrices = currentCryptoPrices.RelatedObject
        .Where(s => groupedInvestments.Any(gi => gi.Symbol == s.Symbol))
        .ToDictionary(s => s.Symbol, p => decimal.Parse(p.PriceUsd));